evap/static/js/datagrid.js

```diff
@@ -0,0 +1,314 @@
+            const button = $(element);
+            const count = this.rows.filter(row => {
+                return row.filterValues.get("state").includes(button.data("filter"));
+            }).length;
+            const badgeClass = count === 0 ? "badge-btn-zero" : "badge-btn";
+            button.append(`<span class="badge badge-pill ${badgeClass}">${count}</span>`);
+
+            button.click(() => {
+                if (button.hasClass("active")) {
+                    button.removeClass("active");
+                    this.filter({state: []});
+                } else {
+                    button.addClass("active");
+                    this.filterButtons.not(button).removeClass("active");
+                    this.filter({state: [button.data("filter")]});
+                }
+            });
+        });
+    }
+
+    addFilterRowData(row, data) {
+        const state = row.find(".icon").get()
+            .map(element => $(element).data("filter"));
+        data.filterValues.set("state", state);
+    }
+
+    get defaultOrder() {
+        return [[5, "asc"]];
+    }
+
+    syncFilterStateWithInputs() {
+        super.syncFilterStateWithInputs();
+        if (this.state.filter.state) {
+            this.filterButtons.filter(`[data-filter="${this.state.filter.state[0]}"]`).addClass("active");
+        }
+    }
+}
+
+// Grid based data grid which has its container separated from its header
+export class ResultGrid extends DataGrid {
+    init({sortableHeaders, filterCheckboxes, sortColumnSelect, sortOrderCheckboxes, resetFilter, resetOrder, ...options}) {
+        this.sortableHeaders = new Map();
+        sortableHeaders.each((index, header) => {
+            const column = $(header).data("col");
+            this.sortableHeaders.set(column, $(header));
+        });
+        this.filterCheckboxes = filterCheckboxes;
+        this.sortColumnSelect = sortColumnSelect;
+        this.sortOrderCheckboxes = sortOrderCheckboxes;
+        this.resetFilter = resetFilter;
+        this.resetOrder = resetOrder;
+        super.init(options);
+    }
+
+    bindEvents() {
+        super.bindEvents();
+        for (const [name, {checkboxes}] of Object.entries(this.filterCheckboxes)) {
+            checkboxes.on("change", () => {
+                const values = checkboxes.filter(":checked").get().map(elem => elem.value);
+                this.filter({[name]: values});
+            });
+        }
+
+        this.sortColumnSelect.add(this.sortOrderCheckboxes).change(() => {
+            const column = this.sortColumnSelect.prop("value");
+            const order = this.sortOrderCheckboxes.filter(":checked").prop("value");
+            if (column === "name-semester") {
+                this.sort([["name", order], ["semester", order]]);
+            } else {
+                this.sort([[column, order]]);
+            }
+        });
+
+        this.resetFilter.click(() => {
+            this.state.filter = {};
+            this.search("");
```

Comment:
:)
I was about to suggest to inline `filter` and `search`, and for those it would work fine, but then `sort` should probably also be inlined and that one is called in three places...

Reply:
Actually, the `filter` method couldn't suffice all use-cases of filtering (reset, update one filter, clear one filter) and it is not really logical to have a separate `search` method.

Currently, everything related to filtering / searching is inlined, which may not be quite what we want. As in my empirical profilings the updates to the DOM are outweighing the filter / sort logic, I could imagine to call `filterRows` and `sortRows` in `update`.
That way, only one call would be needed at the expense of doing potentially unneeded work. What do you think?